keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,450 @@
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. Authorization will be provided by unix standard permissions to the socket file. |
+| Exposing the API to all workloads on the node | Medium | Exposed data via the API is limited to readiness information only. |
+| Kube-API is down or unreachable | Low | Kube-API availability should not impact this API. When the control plane is down or unreachable but Kubelet is working properly this API should return most recent data about local Pods readiness that are computed by Kubelet even if those data were not reported or accepted by Kube-API. |
+
+#### Control Plane availability issue
+
+The Kubernetes Control Plane is the source of truth and guarantees system
+integrity. However, already running Pods should not be dependent on Control
+Plane availability or connectivity issues. When a Node is healthy but has
+problems accessing the Kubernetes API, it should not result in node-local
+components losing track of the health of other node-local components. For
+example, networking components may unnecessarily route to unhealthy endpoints
+if they're relying on stale information from the API Server.
+
+The new API should return Pods readiness that is computed by Kubelet even if
+they are not reported back to kube-API but reflect the current Pod’s state
+locally.
+
+It is worth noting that a node may lose connectivity to the Control Plane even
+if the Control Plane is healthy. In that case, the Node itself would eventually
+be marked as unhealthy. We believe it is still acceptable for node-local
+components to get node-local Pod readiness information as this would not have
+any impact on other Nodes in the cluster.
+
+#### Kubelet restarts issue
+
+There is a Kubelet restart
+[issue](https://github.com/kubernetes/kubernetes/issues/100277) that could
+impact this API. After restart Kubelet marks all Pods as NotReady and then
+starts the init loop checking the actual pod's status. Kubelet should return
+the pod's readiness information only after the initial loop is finished
+otherwise the API should return gRPC error `FAILED_PRECONDITION`. This
+API needs to be reliable and returned data should reflect the actual state of
+the Pods. We don't want this API to return partial data.
+
+## Design Details
+
+### Proposed API
+
+We propose to add new gPRC API `status` in Kubelet, listening on a unix socket at `/var/lib/Kubelet/status/Kubelet.sock`. The endpoint will be versioned.
+The gRPC Service will expose 3 methods serving local Pods statuses data:
+
+```protobuf
+service Status {
+    // ListPodConditions returns a of List of PodConditions
+    rpc ListPodConditions(PodConditionsListRequest) returns (PodConditionsListResponse) {}
+    // WatchPodConditions returns a stream of List of PodConditions
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsListResponse) {}
```

Comment:
how will this be imlpemented? Are we returning individual events when something is changed? Or return all Pods from time to time? Should we have an events-based response instead of the same response type?